module/zfs/spa_errlog.c

```diff
@@ -72,6 +72,11 @@
 
 #define	NAME_MAX_LEN 64
 
+/*
+ * A variable to keep track of check_filesystem() recursion.
+ */
+int zfs_checkfs_recursion = 0;
```

Comment:
This either needs to be locked or better yet placed on the stack and passed as a function argument.  This usage can race when there are multiple pools imported on the same system.

Reply:
Ok, I think I get it, there is a for loop in `zpool_get_errlog()` which iterates over `zfs_ioc_error_log()`, I will try incorporating the `zfs_checkfs_recursion` variable there.

Reply:
This will be tricky because it touches `zfs_ioctl()`. Any thoughts?

Reply:
Then again, the stack tracer reports:
```
 62)     1504     208   check_filesystem+0x487/0x5d0 [zfs]
 63)     1296     208   check_filesystem+0x487/0x5d0 [zfs]
 64)     1088     208   check_filesystem+0x487/0x5d0 [zfs]
 65)      880      88   process_error_block+0xb3/0xc0 [zfs]
 66)      792     128   process_error_log+0x201/0x360 [zfs]
 67)      664      56   spa_get_errlog+0x1cd/0x2a0 [zfs]
 68)      608      40   zfs_ioc_error_log+0x56/0x90 [zfs]
 69)      568     136   zfsdev_ioctl_common+0x662/0x780 [zfs]
 70)      432      32   zfsdev_ioctl+0x53/0xe0 [zfs]
 71)      400      48   __x64_sys_ioctl+0x91/0xd0
 72)      352     176   do_syscall_64+0x5c/0x90
 73)      176     176   entry_SYSCALL_64_after_hwframe+0x63/0xcd
```
I tried with 2 pools and I am able to go into a stack depth of ~43 recursions on *each* pool before it runs out of stack size (same as with just 1 pool with errors imported), suggesting it spawns a new thread for each pool? I am not totally sure.

Reply:
It actually doesn't spawn a new thread, it just goes from pool to pool sequentially.